numpy/core/src/common/get_attr_string.h

```diff
@@ -39,6 +39,14 @@ _is_basic_python_type(PyTypeObject *tp)
     );
 }
 
+static NPY_INLINE void
+clear_exception(PyObject *exception)
+{
+    if (PyErr_Occurred() && PyErr_ExceptionMatches(exception)) {
```

Comment:
Doesn't `PyErr_ExceptionMatches` already contain a check for `PyErr_Occurred`?

Reply:
I'm not sure. Reading the [docs](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_ExceptionMatches), it seemed to me that the check is necessary.

Reply:
You're right, the docs are pretty clear about that.

Reply:
Ah, but tp_getattr isn't allowed to return null without setting an error, is it?

Reply:
well, `tp_getattr`seems to be deprecated, but [`tp_getattro`](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_getattro)  is by default set to [`PyObject_GenericGetAttr()`](https://docs.python.org/3/c-api/object.html#c.PyObject_GenericGetAttr), which indeed raises an `AttributeError` on failure (but [`PyObject_GetAttr()`](https://docs.python.org/3/c-api/object.html#c.PyObject_GetAttr) does not mention raising). Without the `PyErr_Occurred` check the tests don't seem to fail, but that may be risky, especially if a custom function was put there (though I don't know how much of a risk). Ultimately it's not my call, I would be fine with either

Reply:
Doesn't look risky to me. _"`NULL` on failure`"_ means "with an exception set". If an exception weren't set, that would be noted explicitly (as it's very uncommon).

Reply:
So I should remove the check?